src/main/java/com/imsweb/datagenerator/provider/ProviderDataGenerator.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2018 Information Management Services, Inc.
+ */
+package com.imsweb.datagenerator.provider;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.imsweb.datagenerator.DataGenerator;
+
+public abstract class ProviderDataGenerator implements DataGenerator {
+
+    // list of rules to be executed
+    protected List<ProviderDataGeneratorRule> _rules;
+
+    public ProviderDataGenerator() {
+        _rules = new ArrayList<>();
+    }
+
+    @Override
+    public abstract String getId();
+
+    /**
+     * Returns all the provider rules from this generator.
+     */
+    public List<ProviderDataGeneratorRule> getRules() {
+        return _rules;
+    }
+
+    /**
+     * Generates a single provider.
+     * <br/><br/>
+     * @param options options
+     * @return generated provider
+     */
+    public Map<String, String> generateProvider(ProviderDataGeneratorOptions options) {
+        // make sure options are never null
+        if (options == null)
+            options = new ProviderDataGeneratorOptions();
+
+        if ((options.getState() == null) || (options.getState().equals("")))
+            options.setState("MD");
+
+        // execute the facility rules once
+        Map<String, String> provider = new HashMap<>();
+        for (ProviderDataGeneratorRule rule : _rules)
+            rule.execute(provider, options);
+
+        return provider;
+    }
+
+    /**
+     * Generates a list of providers.
+     * <br/><br/>
+     * @param numProviders the number of providers to create
+     * @param options options
+     * @return list of generated providers
+     */
+    public List<Map<String, String>> generateProviders(int numProviders, ProviderDataGeneratorOptions options) {
+        // make sure options are never null
+        if (options == null)
+            options = new ProviderDataGeneratorOptions();
+
+        List<Map<String, String>> providers = new ArrayList<>();
+        for (int i = 0; i < numProviders; i++) {
+            Map<String, String> provider = generateProvider(options);
+            if (!providers.contains(provider))
```

Comment:
It's a bit dicey to rely on Map equality in this case. One would argue that we shouldn't use maps; we should use a Physician DTO and a Facility DTO (and I would agree with that but I am not asking you to change this). With the DTOs, you would be force to define a proper equality/hash. But since you are using the maps, you rely on the map equality/hash, which uses all the key/values pairs.

So if you end-goal is for example to not have two facilities with the same name, this is not going to work well because this code won't check only the name, it will also check all the other properties. And since you generate those randomly, the chances that they are all the same is VERY slim. 

I would suggest you switch to something more explicit. For example, you could check the NPI, or the name, or both. You might also have to move this check into an abstract protected method and implement it in the facility or physician generator.

At the end, the most elegant solution would be to use Facility and Physician DTOs and properly define what uniquely identify them in an equals/hash method. Then this could would be good. It's up to you if you want to make that bigger change, or if you prefer to fix this equality checking.